### PR TITLE
Feat sync followers integration tests

### PR DESCRIPTION
**Pull request by arriqaaq**
- add test suite for synchronous replication
- add test case for basic replication with 2 followers

### CODE REVIEW

**Comment by arriqaaq** [comment]
The tests work fine, but are flaky sometimes with the following errors:

- key not found in one of the followers
- error during follower shutdown

I am thinking of adding this to the immudb-tools repo in case the root cause is not found

**Review by byo — COMMENTED** [review]
Very good approach!

Some more tests I'd see beneficial here:
- turn off one replica and ensure that the master do not commit
- create some commits on the primary first (even without replicas), then turn on replicas, check if those will catch-up
- failover test - start with two replicas, continue committing (maybe in some goroutine), kill one replica but add third one - check if the cluster recovers from that situation
- fault tolerance test - start with 3 replicas and 2 required confirmations, kill single replica, ensure writes are still happening, kill second replica, ensure writes are blocked, bring back one of replicas (it would be more interesting to bring back the first one that was down for longer time) - check if cluster recovers

File: pkg/integration/replication/synchronous_replication_test.go [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+
+// this function executes before the test suite begins execution
+func (suite *SynchronousReplicationTestSuite) SetupSuite() {
+	suite.followers = make([]*server.ImmuServer, 0)
+	suite.followersC = make([]client.ImmuClient, 0)
+	suite.dirs = make([]string, 0)
+
+	//init master server
+	masterDir, err := ioutil.TempDir("", "master-data")
+	assert.NoError(suite.T(), err)
```
[the remark this answers is not in the document]
Comment (arriqaaq) [reply pkg/integration/replication/synchronous_replication_test.go:47]: ack, I will do that

File: pkg/integration/replication/synchronous_replication_test.go [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+}
+
+// this function executes after all tests executed
+func (suite *SynchronousReplicationTestSuite) TearDownSuite() {
+	// stop master
+	if suite.master != nil {
+		suite.master.Stop()
+	}
+
+	time.Sleep(1 * time.Second)
```
[the remark this answers is not in the document]
Comment (arriqaaq) [reply pkg/integration/replication/synchronous_replication_test.go:193]: So there is some issue in creating the bufcon server, that sometimes tests fail during shutdown when closed one after the other, so added some timeout to fix the issue, but I am debugging it further

File: pkg/integration/replication/synchronous_replication_test.go [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+// this function executes after each test case
+func (suite *SynchronousReplicationTestSuite) TearDownTest() {
+	fmt.Println("-- From TearDownTest")
+}
+
+func (suite *SynchronousReplicationTestSuite) TestSyncFromMasterToAllFollowers() {
+	mlr, err := suite.masterC.Login(context.TODO(), []byte(`immudb`), []byte(`immudb`))
+	assert.NoError(suite.T(), err)
+
+	mmd := metadata.Pairs("authorization", mlr.Token)
```
[the remark this answers is not in the document]
Comment (arriqaaq) [reply pkg/integration/replication/synchronous_replication_test.go:223]: ack, thanks for the pointer

File: pkg/integration/replication/synchronous_replication_test.go [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+
+	// restart master
+	opts := server.DefaultOptions().
+		WithMetricsServer(false).
+		WithWebServer(false).
+		WithPgsqlServer(false).
+		WithPort(oldopts.Port).
+		WithDir(oldopts.Dir)
+
+	masterServer := server.DefaultServer().WithOptions(opts).(*server.ImmuServer)
```
[the remark this answers is not in the document]
Comment (arriqaaq) [reply pkg/integration/replication/synchronous_replication_test.go:275]: ack

File: pkg/integration/replication/synchronous_replication_test.go [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+		assert.NoError(suite.T(), err)
+		assert.NotNil(suite.T(), fdb)
+
+		fmd = metadata.Pairs("authorization", fdb.Token)
+		fctx = metadata.NewOutgoingContext(context.Background(), fmd)
+
+		_, err = f.Get(fctx, []byte("key1"))
+		assert.NoError(suite.T(), err)
+
+		_, err = f.Get(fctx, []byte("key2"))
```
[the remark this answers is not in the document]
Comment (arriqaaq) [reply pkg/integration/replication/synchronous_replication_test.go:256]: got it, lemme check how to verify that

File: pkg/integration/replication/synchronous_replication_test.go [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+	}()
+	suite.master = masterServer
+
+	masterPort := masterServer.Listener.Addr().(*net.TCPAddr).Port
+	masterClient, err := client.NewImmuClient(client.DefaultOptions().WithPort(masterPort))
+	assert.NoError(suite.T(), err)
+	assert.NotNil(suite.T(), masterClient)
+	suite.masterC = masterClient
+
+	// refresh followers with updated master port
```
[the remark this answers is not in the document]
Comment (arriqaaq) [reply pkg/integration/replication/synchronous_replication_test.go:291]: This should not be required, lemme confirm why I had added this

File: pkg/integration/replication/synchronous_replication_test.go [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+			MasterPort:       uint32(masterPort),
+			FollowerUsername: "follower",
+			FollowerPassword: "follower1Pwd!",
+		})
+		assert.NoError(suite.T(), err)
+	}
+}
+
+func (suite *SynchronousReplicationTestSuite) TestMasterRestart() {
+	suite.restartMaster()
```
[the remark this answers is not in the document]
Comment (arriqaaq) [reply pkg/integration/replication/synchronous_replication_test.go:312]: ack

File: pkg/integration/replication/synchronous_replication_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (arriqaaq) [reply pkg/integration/replication/synchronous_replication_test.go]: done

File: pkg/integration/replication/synchronous_replication_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (arriqaaq) [reply pkg/integration/replication/synchronous_replication_test.go]: removed

File: pkg/integration/replication/synchronous_replication_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (arriqaaq) [reply pkg/integration/replication/synchronous_replication_test.go]: done

File: pkg/integration/replication/synchronous_replication_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (arriqaaq) [reply pkg/integration/replication/synchronous_replication_test.go]: for some reason post restart on the same port, the clients keep showing invalid username and password unless updated

**Comment by byo** [comment]
@arriqaaq: I'm thinking about one possibility in those integration tests. Currently we're targeting internal immudb instance. However with some good interfaces we could also use exactly the same code to run exactly the same set of actions on external immudb instances (e.g. run through docker containers). WDYT?

I'll try create some quick PoC based on this branch.

**Comment by arriqaaq** [comment]
@byo I agree with you, that would be nice, this would then classify as end to end tests for us

File: pkg/integration/replication/synchronous_replication_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (byo) [reply pkg/integration/replication/synchronous_replication_test.go]: Added in 3d60ece5ff16b02d7570e812f2dd3ae47588583c

**Pull request merged**